src/ax/saga/workflow.go

```diff
@@ -14,75 +15,122 @@ type Workflow struct {
+// and produce commands.
 //
 // It accepts a prototype data instance which is cloned for new instances.
 //
-// For each event type to be handled, the aggregate must implement a "handler"
+// For each message type to be handled, the aggregate must implement a "handler"
 // method that adheres to one of the following signatures:
 //
-//     func (ev *<T>, ax.CommandExecutor)
-//     func (ev *<T>, env ax.Envelope, ax.CommandExecutor)
+//     func (m *<T>, ax.CommandExecutor)
+//     func (m *<T>, env ax.Envelope, ax.CommandExecutor)
 //
-// Where T is a struct type that implements ax.Event.
+// Where T is a struct type that implements ax.Message.
 //
 // Handler methods are responsible for mutating the state of the workflow and
-// producing new commands, based on the event being handled. They may inspect
-// the current state of the workflow, and then return zero or more commands
-// to be executed.
+// producing new commands, based on the message being handled.
 //
-// The names of handler methods are meaningful to the workflow system. If an
-// event is meant to trigger a new workflow instance, its handler method's name
-// must begin with "StartWhen". Other handler methods must begin with "When". By
-// convention these prefixes are followed by the message name, such as:
+// The names of handler methods are meaningful to the workflow system. If a
+// command or event is meant to trigger a new workflow instance, its handler
+// method's name must have a prefix "Begin". Non-triggering command handler
+// methods must have prefix "Do". Both triggering and non-triggering event
+// handler methods must have prefix "When". For triggering event handlers the
+// combined prefix results in "BeginWhen". By convention these prefixes are
+// followed by the message name, such as:
 //
-//     func (*BankTransferWorkflow) StartWhenTransferStarted(*messages.TransferStarted, ax.CommandExecutor)
-//     func (*BankTransferWorkflow) WhenAccountDebited(*messages.AccountDebited, ax.CommandExecutor)
+//      // workflow-triggering command handler
+//      func (*BankTransferWorkflow) BeginDebitAccount(
+//              *messages.DebitAccount,
+//              ax.CommandExecutor,
+//          )
+//
+//      // non-triggering command handler
+//      func (*BankTransferWorkflow) DoDebitAccount(
+//              *messages.DebitAccount,
+//              ax.CommandExecutor,
+//      )
+//
+//      // workflow-triggering event handler
+//      func (*BankTransferWorkflow) BeginWhenAccountDebited(
+//              *messages.AccountDebited,
+//              ax.CommandExecutor,
+//      )
+//
+//      // non-triggering event handler
+//      func (*BankTransferWorkflow) WhenAccountDebited(
+//              *messages.AccountDebited,
+//              ax.CommandExecutor,
+//      )
 func NewWorkflow(p Data) *Workflow {
 	w := &Workflow{
 		Prototype: p,
 	}
 
+	// setup type-switch for command handlers.
+	csw, ctypes, err := typeswitch.New(
```

Comment:
I can't remember... was there a reason we could 't use a single `typeswitch.New()` call with all of the signatures for both commands and events?

If not, we should do this, because it improves two things:

- removes the command/event `switch` from `HandleMessage()`
- prevents duplicate handlers for messages that are both events and commands (which is discouraged but allowed)

Reply:
As discussed, using two separate switches for commands and events is going to stay for now until a future version of switch (which is going to be a separate library) supports adding multiple cases to a single switch instance.